File: Week1/MathProblem/Java/CountPrime.java
import java.util.Arrays;

public class CountPrime {

/*
 * Given an integer n, return the number of prime numbers that are strictly less than n.
 * https://leetcode.com/problems/count-primes/
 * T(c) -> O(n*log(logn)), space -> O(n)
 */
    
    static int countPrimes(int n) {
        /* not prime */
        if(n == 0 || n == 1)
        {
            return 0;
        }
        
        /* using sieve of eratosthenes algorithm */
        boolean primeArr[] = new boolean[n];
        
        Arrays.fill(primeArr, true);
        
        /* 0 and 1 are not prime, marking that as false */
        primeArr[0] = false;
        primeArr[1] = false;
        
        /* running loop till sqrt(n) */
        for(int i=2;i*i<=n;i++)
        {
            /* if number is prime */
            if(primeArr[i])
            {
                /* mark all multiples of i as not prime(false) till n-1 */ 
                for(int j=i*i;j<n;j+=i)
                {
                    primeArr[j] = false;
                }
            }
        }
        
        int count = 0;
        /* counting all prime numbers less than n */
        for(int i=2;i<n;i++)
        {
            if(primeArr[i])
            {
                count++;
            }
        }
        
        return count;
    }

    public static void main(String[] args) {
        int n = 10;
        System.out.println("Primes count less than "+ n +" is = " + countPrimes(n));
        n = 1;
        System.out.println("Primes count less than "+ n +" is = " + countPrimes(n));
    }
}

File: Week1/MathProblem/Python/CountPrime.py
"""
Given an integer n, return the number of prime numbers that are strictly less than n.

https://leetcode.com/problems/count-primes/
"""

def countPrimes(n):
    """ T(c) -> O(n * log(logn)) , S(c) -> O(n) """
    if n<=1:
        return 0
    
    # using sieve of eratosthenes algo
    prime = [True] * (n+1)
    prime[0] = False
    prime[1] = False
    
    i = 2
    while i*i <= n:
        # if number is prime
        if prime[i]:
            # mark multiple of prime(i) as nonprime
            j = i*i
            while j <= n:
                prime[j] = False
                j += i

        i += 1
    
    prime_count = 0
    for p in range(2, len(prime)-1):
        # if prime, count that
        if prime[p]:
            prime_count += 1
        
    return prime_count


n = 10
# There are 4 prime numbers less than 10, they are 2, 3, 5, 7.
print(countPrimes(n))
